chore(core): Track workflow crashes

nikhilkuria opened the pull request:
## Summary

When an instance recovers after a workflow crash, we track the number of executions affected using the pulse telemetry event.

To test this locally 

- Temporarily remove the prod check [here](https://github.com/n8n-io/n8n/blob/master/packages/cli/src/crash-journal.ts#L24) 
- Reduce the frequency of the pulse event [here](https://github.com/n8n-io/n8n/blob/master/packages/cli/src/telemetry/index.ts#L91)
- Create a new source in rudderstack and add the key [here](https://github.com/n8n-io/n8n/blob/master/packages/%40n8n/config/src/configs/diagnostics.config.ts#L22)
- Set up a triggered workflow that runs every second (to ensure we have a bunch of workflows in queue) 
- When you have a few workflows in queue, kill the local server
- create the file `crash.journal` in your local `.n8n` directory 
- start the server again. 
- You should see the crashed workflows metric in rudderstack
<!--
Describe what the PR does and how to test.
Photos and videos are recommended.
-->

## Related Linear tickets, Github issues, and Community forum posts
https://linear.app/n8n/issue/ADO-3815/fix-workflow-error-rate-metric
<!--
Include links to **Linear ticket** or Github issue or Community forum post.
Important in order to close *automatically* and provide context to reviewers.
https://linear.app/n8n/issue/
-->
<!-- Use "closes #<issue-number>", "fixes #<issue-number>", or "resolves #<issue-number>" to automatically close issues when the PR is merged. -->


## Review / Merge checklist

- [ ] PR title and summary are descriptive. ([conventions](../blob/master/.github/pull_request_title_conventions.md)) <!--
   **Remember, the title automatically goes into the changelog.
   Use `(no-changelog)` otherwise.**
-->
- [ ] [Docs updated](https://github.com/n8n-io/n8n-docs) or follow-up ticket created.
- [ ] Tests included. <!--
   A bug is not considered fixed, unless a test is added to prevent it from happening again.
   A feature is not complete without tests.
-->
- [ ] PR Labeled with `release/backport` (if the PR is an urgent fix that needs to be backported)

nikhilkuria committed:
feat(ado-3815): track workflow crashes

nikhilkuria committed:
feat(ado-3815): add missing mock param

nikhilkuria committed:
Merge branch 'master' into feat-ADO-3815-track-workflow-crashes

nikhilkuria committed:
Merge branch 'master' into feat-ADO-3815-track-workflow-crashes

ivov reviewed (changes requested):
@nikhilkuria Can I challenge you a bit? :)

1. Curious why we're including this in the pulse event? The values for crashes in the pulse will stay static until the next restart, unlike all other values that can fluctuate an runtime. Missing context here as the Linear story is empty.
2. This is now calling `Telemetry.trackWorkflowExecution` which tracks `'Workflow execution errored'` for unsuccessful manual executions. Two concerns with this:
	1. `trackWorkflowExecution` on `master` is only ever called from `TelemetryEventRelay.workflowPostExecute` which adds a large number of details to the payload before calling `trackWorkflowExecution`. So by calling it directly now, we are bypassing all these details that are typically sent in the payload for this event. Can this affect the integrity of telemetry data?
	2. During recovery we fire the `workflowExecuteAfter` hook which uses the `EventService` to emit an event that `TelemetryEventRelay` handles by calling `Telemetry.trackWorkflowExecution`. But we already called this before recovery. Is there a risk of double-counting?
3. Unfinished executions do not necessarily equal crashed executions. See `ExecutionRecoveryService`. Unfinished executions may have crashed, but the final determination can only come after we attempt recovery.
4. Can we please keep telemetry decoupled from core services like the event bus?
5. Can we please add tests? Telemetry has [nice coverage](https://github.com/n8n-io/n8n/blob/master/packages/cli/src/events/__tests__/telemetry-event-relay.test.ts).
6. Should the PR title be `chore(core)` rather than `feat(core)`? Telemetry is not functionality that users benefit from.

nikhilkuria commented:
@ivov Thanks for the feedback and sorry for the delay in replying to you here. 
1. We are trying to measure how many of our workflows fail due to errors on n8n side. We can then use it to measure the overall quality/stability of n8n. There are some more details in [this notion page](https://www.notion.so/n8n/Quality-n8n-2095b6e0c94f80048090c71b2b947d70?source=copy_link#2095b6e0c94f80e899fac0acf4094eae). I've also added some content to the linear issue. This will just be one of the pieces to measure that number.
2a. In `trackWorkflowExecution`, manual execution failures are sent with all the properties. We don't do that for prod executions. For prod, we have a counter for the number of workflows that fail. But, I see your point, there is always a risk of us sending events with missing data. 
2b. Possibly, yes. Let me double check that. If all the needed properties are available there, then we don't need the  `trackWorkflowExecution` call that I added.
3. Yes, but we are using this event to track the number of workflows that failed due to infra errors. From the code, I don't see the possibility that a recovered workflow resumes execution. Am I wrong? 
4, 5 and 6. Yes

nikhilkuria committed:
Merge branch 'master' into feat-ADO-3815-track-workflow-crashes

nikhilkuria committed:
feat(ado-3815): remove telemetry from event bus

nikhilkuria committed:
feat(ado-3815): add crashed property

nikhilkuria committed:
feat(ado-3815): add tests for the crashed workflow metrics

nikhilkuria committed:
feat(ado-3815): reverting changes in message bus

tomi reviewed (approved):
LGTM 🚀
